File: src/main/java/FP/CleanLazyList/LazyList.java
package FP.CleanLazyList;

import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.function.Predicate;
import java.util.function.Supplier;

/**
 * A functional, singly-linked lazy list.
 * <p>
 * Key properties:
 * <ul>
 *   <li><b>Structural immutability</b> — nodes never change identity once created.</li>
 *   <li><b>Lazy consing</b> — the tail is a {@link Supplier}, so it isn't computed
 *       until something actually asks for it (mirrors Scala's {@code #::}).</li>
 *   <li><b>Thread-safe memoization</b> — each tail is computed at most once,
 *       even under concurrent access, via double-checked locking.</li>
 * </ul>
 */
public sealed interface LazyList<T> {

    // ---- Factories ----

    /**
     * Returns the canonical empty list.
     */
    static <T> LazyList<T> empty() {
        return new Nil<>();
    }

    /**
     * Builds a node from an eager head and a deferred tail.
     * The tail supplier is not invoked until {@link #tail()} is called.
     */
    static <T> LazyList<T> cons(T head, Supplier<LazyList<T>> tail) {
        return new Cons<>(head, tail);
    }

    // ---- Core API ----

    /** Returns the first element. Throws if the list is empty. */
    T head();

    /** Returns the remainder of the list, forcing lazy evaluation if needed. */
    LazyList<T> tail();

    /** True only for the {@link Nil} case. */
    default boolean isEmpty() {
        return this instanceof Nil<?>;
    }

    // ---- Operations ----

    /**
     * Lazily filters elements: only the returned head is guaranteed to match;
     * everything past it stays deferred.
     * <p>
     * Non-matching elements are skipped via a plain loop rather than recursion,
     * so filtering a long run of rejected elements is stack-safe.
     */
    default LazyList<T> filter(Predicate<? super T> predicate) {
        LazyList<T> current = this;
        while (current instanceof Cons<T> cell) {
            if (predicate.test(cell.head())) {
                // Found a match: emit it, deferring the rest of the filtering.
                return cons(cell.head(), () -> cell.tail().filter(predicate));
            }
            // No match: advance without recursing, keeping the call stack flat.
            current = cell.tail();
        }
        // No matches left: return the empty list.
        return empty();
    }

    /**
     * Eagerly collects up to the first {@code n} elements into a {@link List}.
     * Forces evaluation of exactly the nodes visited — never the one past the
     * {@code n}th element.
     */
    default List<T> take(int n) {
        if (n <= 0) {
            return List.of();
        }
        var result = new ArrayList<T>(n);
        LazyList<T> current = this;
        while (current instanceof Cons<T> cell) {
            result.add(cell.head());
            if (result.size() == n) {
                break;
            }
            current = cell.tail();
        }
        return List.copyOf(result);
    }

    // ---- Node implementations ----

    /** The empty node; a marker with no data. */
    record Nil<T>() implements LazyList<T> {
        @Override
        public T head() {
            throw new NoSuchElementException("Head on empty LazyList");
        }

        @Override
        public LazyList<T> tail() {
            throw new NoSuchElementException("Tail on empty LazyList");
        }
    }

    /**
     * A non-empty node: an eager head plus a tail that is computed lazily
     * and cached after the first evaluation.
     */
    final class Cons<T> implements LazyList<T> {
        private final T head;

        // Cleared once 'tail' is populated, so the closure (and whatever it
        // captured) becomes eligible for garbage collection.
        private Supplier<LazyList<T>> tailProducer;

        // volatile so a tail computed by one thread is visible to others
        // without needing to acquire the lock in the common (already-computed) case.
        private volatile LazyList<T> tail;

        private Cons(T head, Supplier<LazyList<T>> tailProducer) {
            this.head = head;
            this.tailProducer = Objects.requireNonNull(tailProducer);
        }

        @Override
        public T head() {
            return head;
        }

        // -- needs review --

        /**
         * Computes and caches the tail on first access.
         * Double-checked locking avoids synchronizing on every subsequent call
         * while still guaranteeing the supplier runs exactly once.
         */
        @Override
        public LazyList<T> tail() {
            var result = tail;
            if (result == null) {
                synchronized (this) {
                    result = tail;
                    if (result == null) {
                        result = Objects.requireNonNull(tailProducer.get());
                        tail = result;
                        tailProducer = null; // release captured state for GC
                    }
                }
            }
            return result;
        }
    }
}

File: src/main/kotlin/LazyList.kt
package lazylist

/**
 * Functional LazyList featuring structural immutability,
 * lazy consing, and thread-safe tail memoization.
 */
sealed interface LazyList<out T> {
    val head: T
    val tail: LazyList<T>
    val isEmpty: Boolean

    /** A Singleton representing an empty LazyList node. */
    object Nil : LazyList<Nothing> {
        override val head: Nothing
            get() = throw NoSuchElementException("Head on empty LazyList")
        override val tail: Nothing
            get() = throw NoSuchElementException("Tail on empty LazyList")
        override val isEmpty: Boolean = true
    }

    /** Represents a populated node storing a head value and a deferred tail computation. */
    // cons = construct from head and tail
    class Cons<out T>(
        override val head: T,
        tailProducer: () -> LazyList<T>
    ) : LazyList<T> {
        // Caches tail evaluation after first access.
        override val tail: LazyList<T> by lazy(tailProducer)
        override val isEmpty: Boolean = false
    }
}

/**
 * Infix operator constructing a LazyList node.
 * Mirrors Scala's `#::` cons operator.
 */
infix fun <T> T.cons(tail: () -> LazyList<T>): LazyList<T> = LazyList.Cons(this, tail)

/**
 * Lazily filters elements based on a predicate without evaluating non-matching nodes upfront.
 */
fun <T> LazyList<T>.filter(predicate: (T) -> Boolean): LazyList<T> = when (this) {
    is LazyList.Nil -> LazyList.Nil
    is LazyList.Cons -> if (predicate(head)) head cons { tail.filter(predicate) } else tail.filter(predicate)
}

/**
 * Eagerly materializes the first [n] elements into a standard Kotlin List.
 */
fun <T> LazyList<T>.take(n: Int): List<T> = when {
    n <= 0 || isEmpty -> emptyList()
    else -> listOf(head) + tail.take(n - 1)
}

/**
 * Generates an infinite recursive stream of incremental integers starting at [n].
 */
fun numbers(n: Int): LazyList<Int> = n cons { numbers(n + 1) }

/**
 * Recursive prime sieve implementation.
 */
fun primes(numbers: LazyList<Int>): LazyList<Int> =
    numbers.head cons { primes(numbers.tail.filter { it % numbers.head != 0 }) }

fun main() {
    val first10Primes = primes(numbers(2)).take(10)
    println("First 10 primes: $first10Primes")
    // Output: First 10 primes: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
}
